drivers/usbhost/usbhost_hub.c

```diff
@@ -1491,7 +1521,7 @@ static int usbhost_disconnected(struct usbhost_class_s *hubclass)
 {
```

Comment:
can we pass no_delay as a parameter to usbhost_disconnected directly?

Reply:
Do you mean changing  int usbhost_disconnected(struct usbhost_class_s *hubclass) to  int usbhost_disconnected(struct usbhost_class_s *hubclass, bool no_delay) ?

Reply:
Yes

Reply:
If so, can we call usbhost_disconnect_event directly instead setting no_delay flag.

Reply:
@xiaoxiang781216  I think agree with you now and will not use the flag. I am beginning to question my approach. Doing this :
CLASS_DISCONNECTED(child->devclass) 
does not free the other device types quickly. Only hubs are freed immediately. I have to do it your way:
CLASS_DISCONNECTED(child->devclass, true) 
to make it clean. But all the other drivers will have to be modified so ... I am thinking of finding another way to fix the problem. I have done some testing and so far only hubs crash the system when their parent is freed first. After I find out why, I will decide how to fix this.

Reply:
The following line destroys the control pipe and sets hport->ep0 to NULL. The parent runs this.

      _/* Free any resources used by the hub port */_
      _usbhost_hport_deactivate(child);_

When the child calls _usbhost_hubpwr_ the NULL control pipe causes the crash. The original "no_delay" fix works by forcing the child to run before the parent destroys the pipe.
Because there is no need to control power on a removed hub I will push a new fix that deletes that line. It should be safe for the children to disconnect after the parent is gone as long as they don't try to use control pipes.

Reply:
Is this correct for powered hubs?  They will be disconnected, but will remain powered in this case.  Should instead we place a disconnect on the work queue and then have the work item traverse the tree starting from leafs and working back to the root of the disconnect?

Reply:
@ldube how about this suggestion?

Reply:
There is no need to call usbhost_hubpwr because powered hubs with a disconnected USB cable do not receive USB packets. That is why I removed the call. Remember that the driver is running but the hardware has been removed already.